### PR TITLE
Vulkan: Use timeline semaphores for synchronization if possible

### PR DESCRIPTION
**Pull request by msat-huawei**
## What does this PR do?

With this PR the Vulkan RHI backend uses [timeline semaphores](https://www.khronos.org/blog/vulkan-timeline-semaphores) instead of binary semaphores to synchronize execution across queue. Binary semaphores are still used for [swapchains](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkQueuePresentKHR.html#VUID-vkQueuePresentKHR-pWaitSemaphores-03267).
The synchronization still supports binary semaphores for synchronization if timeline semaphores are not available for a device.

### What does the PR actually do?
1. Use timeline semaphores where possible (everywhere but swapchains)
2. Add a new synchronization construct. Binary semaphores have the limitations that the wait operation must be submitted to the queue before the signal operation is. Timeline semaphores don't have this limitation, but when mixing timeline semaphores with binary semaphores we still need to synchronize the queue operations. When enqueuing a binary semaphore wait operation, all dependent semaphore (including timeline semaphores) wait operations must have been submitted already. See [here](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkQueuePresentKHR.html#VUID-vkQueuePresentKHR-pWaitSemaphores-03268) and [here](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkQueueSubmit.html#VUID-vkQueueSubmit-pWaitSemaphores-03238). For binary semaphores this was accomplished by a Signal/Wait on the CPU that ensured the right submit order. We no longer call this functions. Instead we introduce a `SemaphoreTracker` class that counts how many semaphores are on the framegraph before a swapchain, and how many semaphores have already been submitted to the queue. We don't track which specific semaphores have been signaled, just how many semaphores a swapchain depends on have been signaled.
3. A way to tell the framegraph that an external (Fence) semaphore has been signalled. This is accomplished by the `FenceTracker` class. The user of a Fence is responsible to call the `FenceTracker::SignalUserSemaphore` when calling the `Fence::SignalOnCpu` function.

### Why do we need this?

In a previous PR (https://github.com/o3de/o3de/pull/17569) we introduced timeline semaphores for user defined Fences. We found that signalling these from the CPU lead to a deadlock somewhere in the graphics driver/kernel, which is explained by the Vulkan specification linked above.
Signalling Fences from the CPU will be used in MultiGPU contexts. See [this RFC](https://github.com/o3de/sig-graphics-audio/blob/main/rfcs/MultiDeviceSupport/MultiDeviceResources.md)

In addition to fixing the problem above, this may also lead to a slight performance gain, when semaphores are used for syncing between queues. The queue submits no longer need to wait for each other, except for the swapchain.

## How was this PR tested?

Tested with multiple levels and atom sample viewer samples (async compute, multi pipeline, multi scene, mesh) on Windows with Vulkan.
I cannot test this on a machine without timeline semaphore support because I don't have access to hardware. I did deactivate the `m_signalFenceFromCPU` manually however to test this codepath.
This was also tested with the MultiGPU RPI sample in atom sampler viewer in this branch: https://github.com/jhmueller-huawei/o3de/tree/multi-device-copy-pass-test

### CODE REVIEW

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/CommandQueue.cpp
Comment (jhmueller-huawei) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/CommandQueue.cpp:119]: TODO

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuter.cpp
Comment (jhmueller-huawei) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuter.cpp:215]: Leftover TODO

**Review by jhmueller-huawei — APPROVED** [review]
Looks good, great job! Just minor comments from my side, but I'd like to get this reviewed by @moudgils as well.

**Comment by moudgils** [comment]
Recommend waiting for @akioCL approval before merging. I will also try to find time to go through it this week.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/CommandQueue.cpp
Comment (msat-huawei) [reply Gems/Atom/RHI/Vulkan/Code/Source/RHI/CommandQueue.cpp]: Fixed this. This would have been a problem with more than one Fence in a Queue submit

**Comment by msat-huawei** [comment]
> I'm going through the review but I feel like this change increased the complexity of semaphores a lot. This is starting to look like spaghetti code. Maybe this is a signal that a deeper refactor is need in order to accommodate binary and timeline semaphores. For example, almost all functions of the semaphore class have a different behavior for binary and timeline semaphore. Maybe we should separate into two classes with a common interface? I'm having a hard time following the semaphore logic in the Vulkan RHI. What do you guys think?

I refactored the code a bit so I could remove the FenceTracker class from the public RHI API. The `SemaphoreTracker` classes are not signalled and waited-for through the `Fence`/`Semaphore` classes. This removes a lot of the new code I added.

Do you mean the implementation in the `Semaphore`/`Fence` with Spaghetti code? Should I split them into a timeline semaphore and binary semaphore/VkFence specific class?

**Comment by msat-huawei** [comment]
I'm gettings some validation errors like:
```
vkDebugMessage: [ERROR][Validation] Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object 0: handle = 0x20582dd94b0, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x4dae5635 | vkQueueSubmit(): pSubmits[0].pCommandBuffers[0] command buffer VkCommandBuffer 0x20582dd94b0[] expects VkImage 0x9484f20000012180[LuminanceMapAttachmentId] (subresource: aspectMask 0x1 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL.
vkDebugMessage: [ERROR][Validation] Validation Error: [ UNASSIGNED-VkImageMemoryBarrier-image-00004 ] Object 0: handle = 0x20582dd94b0, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x4acfa767 | vkQueueSubmit(): in submitted command buffer VkImageMemoryBarrier acquiring ownership of VkImage (VkImage 0x35807000000d02e[]), from srcQueueFamilyIndex 0 to dstQueueFamilyIndex 2 has no matching release barrier queued for execution.
[ERROR][Validation] Validation Error: [ UNASSIGNED-VkImageMemoryBarrier-image-00004 ] Object 0: handle = 0x2060a142060, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x4acfa767 | vkQueueSubmit(): in submitted command buffer VkImageMemoryBarrier acquiring ownership of VkImage (VkImage 0x35807000000d02e[]), from srcQueueFamilyIndex 2 to dstQueueFamilyIndex 0 has no matching release barrier queued for execution.
```
This are likely false positives as discussed [here](https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/2441) and [here](https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/4427).

After the refactoring the PR does not longer work for the MultiGPU sample in [this](https://github.com/jhmueller-huawei/o3de/tree/multi-device-copy-pass-test) branch. I will mark the PR as draft until I fixed that

**Comment by msat-huawei** [comment]
I don't think we have a problem when only binary semaphores are present.
When we submit a wait operation for a semaphore we wait for the signal operation to finish.
Before the signal operation is submitted, we wait for the signal operations of the directly dependent semaphores to be enqueued.
The dependent semaphores wait on their dependencies and so on.
Or maybe I don't understand something there.

I will split the Semaphore/Fence classes into normal/timeline semaphore classes.

Your approach also makes sense. Moving it to the FrameGraphCompiler also makes sense. I will implement it.

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphCompiler.cpp
Comment (martinwinter-huawei) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphCompiler.cpp:1]: Please revert formatting changes

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuteGroupSecondary.h
Comment (martinwinter-huawei) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuteGroupSecondary.h:1]: Formatting change?

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuteGroupSecondary.cpp
Comment (martinwinter-huawei) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuteGroupSecondary.cpp:1]: Formatting change?

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuteGroupPrimary.cpp
Comment (martinwinter-huawei) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuteGroupPrimary.cpp:1]: Formatting change?

File: Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuter.cpp
Comment (martinwinter-huawei) [inline Gems/Atom/RHI/Vulkan/Code/Source/RHI/FrameGraphExecuter.cpp:1]: Unfortunately also a lot of formatting

**Review by martinwinter-huawei — APPROVED** [review]

**Comment by msat-huawei** [comment]
@akioCL I refactored the synchronization code according to your suggestion, and split the Fence/Semaphore into multiple classes. Can you have a look at it again?

**Comment by martinwinter-huawei** [comment]
@akioCL we did another refactor, what do you think about this?

**Pull request merged**
